File: Perl/ZNR.pl
#!/usr/bin/perl
#This script sends the notes of your zone to a CSV file. Inputting a zone is required. 
#You can set a limit of notes to display and epoch time if preferred.
#The credentials are read in from a configuration file in the same directory.
#The file is named config.cfg in the format:

#[Dynect]
#cn: customer_name
#un: user_name
#pw: password

#Usage: %perl ZNR.pl -z example.com [-l 10] [-e] [-f zone_notes.csv]
#This will print to the file zone_notes.csv to a CSV file with a limit of 10 notes and epoch time.

# Options
# -h --help			Show the help message and exit
# -z --zone			Search for zone report with zone name
# -l --limit	 		The maximum number of notes to retrieve
# -e --epoch			Use epoch time instead of a formatted time
# -f --file			Set file name. Default: notes_[zonename].csv

use warnings;
use strict;
use Config::Simple;
use Getopt::Long;
use LWP::UserAgent;
use JSON;
use Text::CSV;
use POSIX qw( strftime );

#Import DynECT handler
use FindBin;
use lib "$FindBin::Bin/DynECT";  # use the parent directory
require DynECT::DNS_REST;

#Get Options
my $opt_list=0; #Initalized to check if limit is set
my $opt_file=""; #Initalized to see check against opt_file being set
my $opt_zone=""; #Initalized to see check against opt_zone being set
my $opt_help;
my $opt_epoch;
my $notelist;
my %api_param;

GetOptions(
	'help' => \$opt_help,
	'epoch' => \$opt_epoch,
	'limit=i' => \$opt_list,
	'file=s' => \$opt_file,
	'zone=s' =>\$opt_zone,
);

#Printing help menu
if ($opt_help) {
	print "\tOptions:\n";
	print "\t\t-h, --help\t\t Show the help message and exit\n";
	print "\t\t-e, --epoch\t\t Use epoch time instead of a formatted time\n";
	print "\t\t-f, --file\t\t Set file name. Default: notes_[zonename].csv\n";
	print "\t\t-l, --limit\t\t Set the maximum number of notes to retrieve (Newest first)\n";
	print "\t\t-z, --zone\t\t Name of zone (Required)\n\n";
	exit;
}

#Checking if -z is set
elsif ($opt_zone eq "")
{
	print "Zonename needs to be set. Use \"-z [zonename]\"\n";
	exit;
}

#Create config reader
my $cfg = new Config::Simple();

# read configuration file (can fail)
$cfg->read('config.cfg') or die $cfg->error();

#dump config variables into hash for later use
my %configopt = $cfg->vars();
my $apicn = $configopt{'cn'} or do {
	print "Customer Name required in config.cfg for API login\n";
	exit;
};

my $apiun = $configopt{'un'} or do {
	print "User Name required in config.cfg for API login\n";
	exit;
};

my $apipw = $configopt{'pw'} or do {
	print "User password required in config.cfg for API login\n";
	exit;
};

#API login
my $dynect = DynECT::DNS_REST->new;
$dynect->login( $apicn, $apiun, $apipw) or
	die $dynect->message;

# Setting up new csv file
$opt_file = "notes_$opt_zone.csv" unless($opt_file ne ""); #Set filename if -f else, use default
my $fh;
my $csv = Text::CSV->new ( { binary => 1, eol => "\n" } ) or die "Cannot use CSV: ".Text::CSV->error_diag ();
open $fh, ">", $opt_file  or die "new.csv: $!";
print "Writing CSV file to: $opt_file\n";

# Setting header information
$csv->print($fh, [ "User", "Type", "When", "Note"]);


#If -l is set then send then limit. If it is not, assume no limit.
if($opt_list!=0)
	{%api_param = (zone => $opt_zone, limit => $opt_list);}
else
	{%api_param = (zone => $opt_zone);}
$dynect->request( "/REST/ZoneNoteReport", 'POST',  \%api_param) or die $dynect->message;

#Print out the zone, type, time and note to the user/file
foreach my $zoneIn (@{$dynect->result->{'data'}})
{
	my $user = $zoneIn->{'user_name'};
	my $type = $zoneIn->{'type'};
	my $time = $zoneIn->{'timestamp'};
	my $note = $zoneIn->{'note'};
	chomp $note;
	$time = strftime("%b %d, %Y (%H:%M - UTC)", gmtime($time)) unless($opt_epoch); #Set formatted time unless -e
	$csv->print ($fh, [ $user, $type, $time, $note ] );
}

# Close csv file
close $fh or die "$!";
print "CSV file write sucessful.\n";

#API logout
$dynect->logout;

File: PHP/znr.php
#!/usr/bin/php
<?php
#!/usr/bin/perl
#This script sends the notes of your zone to a CSV file. Inputting a zone is required. 
#You can set a limit of notes to display and epoch time if preferred.
#The credentials are read in from a configuration file in the same directory.
#The file is named config.ini in the format:

#API Login Information
#cn= customer_name
#un= user_name
#pw= password

# Options
# -h		Show the help message and exit
# -z		Search for zone report with zone name
# -l 		The maximum number of notes to retrieve
# -e		Use epoch time instead of a formatted time
# -f		Set file name. Default: notes_[zonename].csv

#Usage: %php znr.php -z example.com [-l 10] [-e] [-f zone_notes.csv]
#This will print to the file zone_notes.csv to a CSV file with a limit of 10 notes and epoch time.

#Get options from command line
$shortopts .= "z:"; 
$shortopts .= "l:";  
$shortopts .= "f:";  
$shortopts .= "e";  
$shortopts .= "h"; 
$options = getopt($shortopts);

$opt_zone .= $options["z"]; 
$opt_limit .= $options["l"]; 
$opt_file .= $options["f"]; 
$opt_epoch .= $options["e"]; 
date_default_timezone_set('UTC'); #Sets timezone to UTC for datetime

#Print help menu
if (is_bool($options["h"])) {
        print "Options:\n";
	print "-h\tShow the help message and exit\n";
	print "-e\tUse epoch time instead of a formatted time\n";
	print "-f\tSet file name. Default: notes_[zonename].csv\n";
	print "-l\tSet the maximum number of notes to retrieve (Newest first)\n";
	print "-z\tName of zone (Required)\n\n";
	exit;}
		
# Parse ini file (can fail)
#Set the values from file to variables or die
$ini_array = parse_ini_file("config.ini") or die;
$api_cn = $ini_array['cn'] or die("Customer Name required in config.ini for API login\n");
$api_un = $ini_array['un'] or die("User Name required in config.ini for API login\n");
$api_pw = $ini_array['pw'] or die("Password required in config.ini for API login\n");	

# Prevent the user from proceeding if they have not entered -z
if($opt_zone == "")
{
	print "You must enter \"-z [example.com]\"\n";
	exit;
}

# Log into DYNECT
# Create an associative array with the required arguments
$api_params = array(
			'customer_name' => $api_cn,
			'user_name' => $api_un,
			'password' => $api_pw);
$session_uri = 'https://api2.dynect.net/REST/Session/'; 
$decoded_result = api_request($session_uri, 'POST', $api_params,  $token);	

#Set the token
if($decoded_result->status == 'success')	
	{$token = $decoded_result->data->token;}

# Setting file name and opening file for writing
if($opt_file == "")
	{$opt_file = "notes_$opt_zone.csv";}
$fp = fopen($opt_file, 'w') or die;
print "Writing CSV file to: $opt_file\n";

#If -l is set then send then limit. If it is not, assume no limit.
if($opt_limit!=0)
        {$api_param = array('zone' => $opt_zone, 'limit' => $opt_limit);}
else
        {$api_param = array('zone' => $opt_zone);}
$session_uri = "https://api2.dynect.net/REST/ZoneNoteReport";
$decoded_result = api_request($session_uri, 'POST', $api_param, $token);

# Go through the result assigning values to results
foreach ($decoded_result->data as $zoneIn)
{
        $user = $zoneIn->user_name;
	$type = $zoneIn->type;
        $note =  $zoneIn->note;
	$note = rtrim($note);
        $time =  $zoneIn->timestamp;
	$dt = new DateTime("@$time"); # Set epoch time to datetime
	#If epoch is set dont format the time
	if(!is_bool($options['e']))
		{$time=  $dt->format('M d, Y (H:i - \U\T\C)');} # Format date
	fputcsv($fp, array($user, $type, $time, $note)); #Send evertying in the array to the csv
}

#Close file and let user know
fclose($fp) or die("Could not close file $opt_file");
print "CSV file write sucessful.";

# Logging Out
$session_uri = 'https://api2.dynect.net/REST/Session/'; 
$api_params = array (''=>'');
$decoded_result = api_request($session_uri, 'DELETE', $api_params,  $token);	

# Function that takes zone uri, request type, parameters, and token.
# Returns the decoded result
function api_request($zone_uri, $req_type, $api_params, $token)
{
	$ch = curl_init();
	curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);  # TRUE to return the transfer as a string of the return value of curl_exec() instead of outputting it out directly.
	curl_setopt($ch, CURLOPT_FAILONERROR, false); # Do not fail silently. We want a response regardless
	curl_setopt($ch, CURLOPT_HEADER, false); # disables the response header and only returns the response body
	curl_setopt($ch, CURLOPT_HTTPHEADER, array('Content-Type: application/json','Auth-Token: '.$token)); # Set the token and the content type so we know the response format
	curl_setopt($ch, CURLOPT_CUSTOMREQUEST, $req_type);
	curl_setopt($ch, CURLOPT_URL, $zone_uri); # Where this action is going,
	curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($api_params));
	
	$http_result = curl_exec($ch);
	$decoded_result = json_decode($http_result); # Decode from JSON as our results are in the same format as our request
	
	if($decoded_result->status != 'success')
		{$decoded_result = api_fail($token, $decoded_result);}  	
	
	return $decoded_result;
}

#Expects 2 variable, first a reference to the API key and second a reference to the decoded JSON response
function api_fail($token, $api_jsonref) 
{
	#loop until the job id comes back as success or program dies
	while ( $api_jsonref->status != 'success' ) {
        	if ($api_jsonref->status != 'incomplete') {
                       foreach($api_jsonref->msgs as $msgref) {
                                print "API Error:\n";
                                print "\tInfo: " . $msgref->INFO . "\n";
                                print "\tLevel: " . $msgref->LVL . "\n";
                                print "\tError Code: " . $msgref->ERR_CD . "\n";
                                print "\tSource: " . $msgref->SOURCE . "\n";
                        };
                        #api logout or fail
			$session_uri = 'https://api2.dynect.net/REST/Session/'; 
			$api_params = array (''=>'');
			$decoded_result = api_request($session_uri, 'DELETE', $api_params,  $token);	
                        exit;
                }
                else {
                        sleep(5);
                        $session_uri = "https://api2.dynect.net/REST/Job/" . $api_jsonref->job_id ."/";
			$api_params = array (''=>'');
			$decoded_result = api_request($session_uri, 'GET', $api_params,  $token);	
               }
        }
        return $api_jsonref;
}
?>


